File: src/bin/io/RandomAccessData.java
package bin.io;

import java.io.Closeable;
import java.io.IOException;

/**
 * @author Bin
 */
public interface RandomAccessData extends Closeable {

    void seek(long pos) throws IOException;

    int read(byte[] data, int off, int len) throws IOException;

    void write(byte[] data, int off, int len) throws IOException;

    long length() throws IOException;

    void setLength(long newLength) throws IOException;

    long position() throws IOException;

    void sync() throws IOException;

    String getName();

    RandomAccessData getAnotherInSameParent(String name) throws IOException;

    RandomAccessData newSameInstance() throws IOException;

    default RandomAccessData newFragment(long offset, long length) throws IOException {
        return new FragmentRandomAccessData(newSameInstance(), offset, length);
    }
}

File: Android/app/src/main/java/bin/io/RandomAccessData.kt
package bin.io

import java.io.Closeable
import java.io.IOException

/**
 * @author Bin
 */
interface RandomAccessData : Closeable {
    @Throws(IOException::class)
    fun seek(pos: Long)

    @Throws(IOException::class)
    fun read(data: ByteArray, off: Int, len: Int): Int

    @Throws(IOException::class)
    fun write(data: ByteArray, off: Int, len: Int)

    @Throws(IOException::class)
    fun length(): Long

    @Throws(IOException::class)
    fun setLength(newLength: Long)

    @Throws(IOException::class)
    fun position(): Long

    @Throws(IOException::class)
    fun sync()
    val name: String

    @Throws(IOException::class)
    fun getAnotherInSameParent(name: String): RandomAccessData

    @Throws(IOException::class)
    fun newSameInstance(): RandomAccessData

    @Throws(IOException::class)
    fun newFragment(offset: Long, length: Long): RandomAccessData {
        return FragmentRandomAccessData(newSameInstance(), offset, length)
    }
}